# 01 – Data Overview

## Ziel
Erster Überblick über die TED-Daten: Struktur, Spalten, Datentypen, grundlegende Statistiken.

## Schritte
- Daten laden
- Spalten anzeigen
- Datentypen prüfen
- Fehlende Werte analysieren
- Erste Hypothesen formulieren


---------------------

## BASE SETTINGS

--------------------------- 

### Import moduls

In [1]:
# import standard modules
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path

import sys
from pathlib import Path

 ### Setup style

In [2]:
# show all columns
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# visualisation settings
pd.set_option('display.float_format', '{:,.2f}'.format)

### Load scripts

In [3]:
%load_ext autoreload
%autoreload 2

from pathlib import Path
import sys

# found min directory
PROJECT_ROOT = Path("..").resolve()

# maindirectory sys.path
if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

# import function from script
from my_scripts.preprocessing import (
    basic_preprocessing,
    filter_germany,
    load_years,
    overview,
)

--------------------

## BASE PREPARATION

------------------------

In [5]:
# ---------------------------------------------------------
# Import dataset
# ---------------------------------------------------------

# set setting
data_path = Path("..\\data")

# select all years
years = list(range(2008, 2017))   

# downloads for all years
df = load_years(data_path, years)

In [ ]:
# ---------------------------------------------------------
Preprocessing
# ---------------------------------------------------------

# change empty values ​​with NaNin target 
df = basic_preprocessing(df)

---------------------

## EDA

---------------

-----------------
### Basic structure

---------------

In [6]:
print("Shape:", df.shape)

Shape: (4362869, 75)


In [7]:
df.head()

,ID_NOTICE_CAN,TED_NOTICE_URL,YEAR,ID_TYPE,DT_DISPATCH,XSD_VERSION,CANCELLED,CORRECTIONS,B_MULTIPLE_CAE,CAE_NAME,CAE_NATIONALID,CAE_ADDRESS,CAE_TOWN,CAE_POSTAL_CODE,CAE_GPA_ANNEX,ISO_COUNTRY_CODE,ISO_COUNTRY_CODE_GPA,B_MULTIPLE_COUNTRY,ISO_COUNTRY_CODE_ALL,CAE_TYPE,EU_INST_CODE,MAIN_ACTIVITY,B_ON_BEHALF,B_INVOLVES_JOINT_PROCUREMENT,B_AWARDED_BY_CENTRAL_BODY,TYPE_OF_CONTRACT,TAL_LOCATION_NUTS,B_FRA_AGREEMENT,FRA_ESTIMATED,B_FRA_CONTRACT,B_DYN_PURCH_SYST,CPV,MAIN_CPV_CODE_GPA,ID_LOT,ADDITIONAL_CPVS,B_GPA,GPA_COVERAGE,LOTS_NUMBER,VALUE_EURO,VALUE_EURO_FIN_1,VALUE_EURO_FIN_2,B_EU_FUNDS,TOP_TYPE,B_ACCELERATED,OUT_OF_DIRECTIVES,CRIT_CODE,CRIT_PRICE_WEIGHT,CRIT_CRITERIA,CRIT_WEIGHTS,B_ELECTRONIC_AUCTION,NUMBER_AWARDS,ID_AWARD,ID_LOT_AWARDED,INFO_ON_NON_AWARD,INFO_UNPUBLISHED,B_AWARDED_TO_A_GROUP,WIN_NAME,WIN_NATIONALID,WIN_ADDRESS,WIN_TOWN,WIN_POSTAL_CODE,WIN_COUNTRY_CODE,B_CONTRACTOR_SME,CONTRACT_NUMBER,TITLE,NUMBER_OFFERS,NUMBER_TENDERS_SME,NUMBER_TENDERS_OTHER_EU,NUMBER_TENDERS_NON_EU,NUMBER_OFFERS_ELECTR,AWARD_EST_VALUE_EURO,AWARD_VALUE_EURO,AWARD_VALUE_EURO_FIN_1,B_SUBCONTRACTED,DT_AWARD
0,200880,ted.europa.eu/udl?uri=TED:NOTICE:80-2008:TEXT:...,2008,3,11-DEC-07,D205,0,0,NaN,St. Elisabeth-Krankenhaus Leipzig,NaN,Biedermannstraße 84,Leipzig,04277,NaN,DE,NaN,NaN,NaN,8,NaN,Health;Social protection,NaN,NaN,NaN,W,DED31,N,NaN,N,NaN,"45,215,000.00",NaN,NaN,45000000,N,NaN,0.00,NaN,"302,964.15","302,964.15",NaN,OPE,NaN,0,M,NaN,Preis---Qualität,60---40,N,1,"2,157,402.00",NaN,NaN,N,NaN,Caverion GmbH Gebäudetechnik,NaN,Walter-Köhn-Straße 1c,Leipzig,04356,DE,NaN,265-02-028,NaN,2.00,NaN,NaN,NaN,NaN,NaN,"302,964.15","302,964.15",NaN,29-SEP-07
1,200881,ted.europa.eu/udl?uri=TED:NOTICE:81-2008:TEXT:...,2008,3,27-DEC-07,D205,0,0,NaN,Stadt Wolfsburg,NaN,Porschestr. 49,Wolfsburg,38440,NaN,DE,NaN,NaN,NaN,3,NaN,General public\services,N,NaN,NaN,W,DE913,N,NaN,N,NaN,"45,215,100.00",NaN,NaN,45110000---45223220---45231300---45261420---45...,Y,NaN,0.00,NaN,"478,780.08","478,780.08",N,OPE,NaN,0,L,100.00,NaN,NaN,N,1,"2,157,403.00",NaN,NaN,N,NaN,Reitmeier Bau und Wohnungsbau GmbH,NaN,Leinering 21,Wolfsburg,38446,DE,NaN,07-0204 / 07-0274,NaN,3.00,NaN,NaN,NaN,NaN,NaN,"478,780.08","478,780.08",N,21-DEC-07
2,200882,ted.europa.eu/udl?uri=TED:NOTICE:82-2008:TEXT:...,2008,3,27-DEC-07,D205,0,0,NaN,Ville de Pau,NaN,Hôtel de Ville - place Royale,Pau,64036,NaN,FR,NaN,NaN,NaN,3,NaN,NaN,N,NaN,NaN,W,NaN,Y,NaN,Y,NaN,"45,421,145.00",NaN,NaN,NaN,Y,NaN,0.00,NaN,NaN,NaN,N,OPE,NaN,0,M,NaN,Prix---Délai d'intervention d'urgence,80---20,N,1,"2,157,404.00",NaN,NaN,N,NaN,Société TEXIA,NaN,15; rue Richelieu,Pau,64000,FR,NaN,2007-136,NaN,1.00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,20-NOV-07
3,200883,ted.europa.eu/udl?uri=TED:NOTICE:83-2008:TEXT:...,2008,3,27-DEC-07,D205,0,0,NaN,Opac de Saint-Quentin,NaN,142; boulevard Gambetta; B.P. 41,Saint-Quentin,02101,NaN,FR,NaN,NaN,NaN,6,NaN,Housing and community amenities,N,NaN,NaN,W,NaN,N,NaN,N,NaN,"45,211,100.00",NaN,NaN,NaN,Y,NaN,0.00,NaN,"2,255,999.00","2,255,999.00",NaN,NIC,NaN,0,M,NaN,prix des prestations---valeur technique de l'o...,55---45,NaN,1,"2,157,405.00",NaN,NaN,N,NaN,S.A.C ( Société anizienne de construction),NaN,NaN,Anizy- le-Château,02320,FR,NaN,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"2,255,999.00","2,255,999.00",NaN,07-DEC-07
4,200884,ted.europa.eu/udl?uri=TED:NOTICE:84-2008:TEXT:...,2008,3,27-DEC-07,D205,0,0,NaN,Ministerio de Fomento; Secretaria de Estado de...,NaN,Plaza de los Sagrados Corazones; 7,Madrid,28071,NaN,ES,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,NaN,W,NaN,N,NaN,N,NaN,"45,234,116.00",NaN,NaN,NaN,NaN,NaN,0.00,NaN,"23,462,327.59","23,462,327.59",NaN,OPE,NaN,0,M,NaN,Oferta Técnica---Oferta económica,70---30,NaN,1,"2,157,406.00",NaN,NaN,N,NaN,Sociedad Anonima de Obras y Servicios; Copasa,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"30,657,786.21","23,462,327.59","23,462,327.59",NaN,14-DEC-07


In [8]:
df.columns

Index(['ID_NOTICE_CAN', 'TED_NOTICE_URL', 'YEAR', 'ID_TYPE', 'DT_DISPATCH',
       'XSD_VERSION', 'CANCELLED', 'CORRECTIONS', 'B_MULTIPLE_CAE', 'CAE_NAME',
       'CAE_NATIONALID', 'CAE_ADDRESS', 'CAE_TOWN', 'CAE_POSTAL_CODE',
       'CAE_GPA_ANNEX', 'ISO_COUNTRY_CODE', 'ISO_COUNTRY_CODE_GPA',
       'B_MULTIPLE_COUNTRY', 'ISO_COUNTRY_CODE_ALL', 'CAE_TYPE',
       'EU_INST_CODE', 'MAIN_ACTIVITY', 'B_ON_BEHALF',
       'B_INVOLVES_JOINT_PROCUREMENT', 'B_AWARDED_BY_CENTRAL_BODY',
       'TYPE_OF_CONTRACT', 'TAL_LOCATION_NUTS', 'B_FRA_AGREEMENT',
       'FRA_ESTIMATED', 'B_FRA_CONTRACT', 'B_DYN_PURCH_SYST', 'CPV',
       'MAIN_CPV_CODE_GPA', 'ID_LOT', 'ADDITIONAL_CPVS', 'B_GPA',
       'GPA_COVERAGE', 'LOTS_NUMBER', 'VALUE_EURO', 'VALUE_EURO_FIN_1',
       'VALUE_EURO_FIN_2', 'B_EU_FUNDS', 'TOP_TYPE', 'B_ACCELERATED',
       'OUT_OF_DIRECTIVES', 'CRIT_CODE', 'CRIT_PRICE_WEIGHT', 'CRIT_CRITERIA',
       'CRIT_WEIGHTS', 'B_ELECTRONIC_AUCTION', 'NUMBER_AWARDS', 'ID_AWARD',
       'ID_LOT_AWA

#### Notes: Basic structure
The dataset contains 4,362,869 rows and 75 columns.
A detailed description of all column meanings, formats, and data definitions is provided in the Data_dictionary file located in the data/ folder.

--------------------

### Data quality

-----------------

In [9]:
overview(df)

,dtype,total,missing_n,missing_%,uniques_n,uniques
ID_NOTICE_CAN,int64,4362869,0,0.00,1404322,"[200880, 200881, 200882, 200883, 200884, 20088..."
TED_NOTICE_URL,object,4362869,0,0.00,1404322,[ted.europa.eu/udl?uri=TED:NOTICE:80-2008:TEXT...
YEAR,int64,4362869,0,0.00,9,"[2008, 2009, 2010, 2011, 2012, 2013, 2014, 201..."
ID_TYPE,int64,4362869,0,0.00,7,"[3, 6, 18, 25, 21, 23, 22]"
DT_DISPATCH,object,4362869,0,0.00,3297,"[11-DEC-07, 27-DEC-07, 28-DEC-07, 19-DEC-07, 2..."
XSD_VERSION,object,4362869,0,0.00,7,"[D205, D206, D207, R207.S3, R208.S1, R208.S2, ..."
CANCELLED,int64,4362869,0,0.00,2,"[0, 1]"
CORRECTIONS,int64,4362869,0,0.00,10,"[0, 1, 2, 4, 3, 5, 84, 7, 8, 6]"
B_MULTIPLE_CAE,object,148895,4213974,96.59,2,"[nan, N, Y]"
CAE_NAME,object,4362867,2,0.00,196676,"[St. Elisabeth-Krankenhaus Leipzig, Stadt Wolf..."


#### Notes: Data Quality Overview

1. Columns with excessive missing data (recommended for removal)
Several variables contain extremely high proportions of missing values (often >90–95%), making them unsuitable for analysis or modeling. These include:
- WIN_NAME, WIN_ADDRESS, WIN_TOWN, WIN_POSTAL_CODE — award‑stage winner information is missing for the majority of records.
- CRIT_PRICE_WEIGHT, CRIT_QUALITY_WEIGHT — evaluation criteria weights are sparsely populated and inconsistent across years.
- AWARD_VALUE_EURO_FIN_1 — secondary award value field with large gaps and duplication of the main award value.
- B_SUBCONTRACTED — subcontracting indicator is missing for more than half of observations.
- ADDITIONAL_CPVS — supplementary CPV codes appear only in a small subset of tenders.
- GPA_COVERAGE, B_GPA — GPA‑related fields are mostly empty and not consistently reported.
- These variables provide limited analytical value and can be safely excluded from the core dataset.

2. Columns requiring type corrections
Several fields are stored as object but represent structured or numeric information:
- DT_AWARD, DT_DISPATCH, DT_RECEIPT
  - Currently stored as strings.
  - Should be converted to datetime for temporal analysis and modeling.
- VALUE_EURO, AWARD_VALUE_EURO, AWARD_EST_VALUE_EURO
  - Numeric fields with missing values and occasional formatting inconsistencies.
  - Convert to float, ensure coercion of invalid entries, and consider log‑transformation for modeling.
- ISO_COUNTRY_CODE, TYPE_OF_CONTRACT, MAIN_CPV_CODE_GPA
  - Categorical fields stored as object.
  - Convert to category to reduce memory usage and improve model performance.
- NUMBER_OFFERS
  - Should be numeric (int), but contains non‑numeric placeholders.
  - Convert using pd.to_numeric(errors="coerce").

3. Columns with high cardinality (handle with care)
Some variables contain extremely large numbers of unique values:
- TED_NOTICE_URL — unique per tender; not useful for modeling.
- CAE_ADDRESS, CAE_TOWN, CAE_NAME — contracting authority identifiers; high cardinality and inconsistent formatting.
- WIN_NAME, WIN_ADDRESS — winner identifiers; high cardinality and sparse.
These fields are not suitable for direct modeling and should be excluded or replaced with aggregated features (e.g., country‑level or CPV‑level indicators).

4. Columns with consistent and reliable structure (recommended to keep)
These variables are well‑populated and analytically meaningful:
- YEAR — newly added; essential for trend analysis.
- ISO_COUNTRY_CODE — complete and consistent.
- TYPE_OF_CONTRACT — stable categorical variable.
- MAIN_CPV_CODE_GPA — core classification of procurement subject.
- NUMBER_OFFERS — key variable for competition analysis.
- VALUE_EURO / AWARD_VALUE_EURO — central economic indicators.
These should form the backbone of your analytical dataset.

5. Additional recommended preprocessing steps
- Remove exact duplicate rows (your dataset contains >2.5M duplicates).
- Normalize CPV codes (strip whitespace, ensure consistent formatting).
- Standardize country codes (ISO 2‑letter format is already good).
- Create derived features such as:
  - IS_FAILED_TENDER (0–1 offers)
  - IS_LOW_COMPETITION (2 offers)
  - LOG_VALUE_EURO
  - CPV_SECTION (first 2 digits of CPV)
These will significantly improve interpretability and model performance.

-----------------
### Descriptive statistics
---------------

In [10]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
ID_NOTICE_CAN,"4,362,869.00","1,512,230,040.81","818,517,533.27","20,155.00","201,622,617.00","2,011,239,877.00","2,014,229,433.00","2,016,466,898.00"
YEAR,"4,362,869.00","2,012.32",2.52,"2,008.00","2,010.00","2,012.00","2,015.00","2,016.00"
ID_TYPE,"4,362,869.00",3.25,1.14,3.00,3.00,3.00,3.00,25.00
CANCELLED,"4,362,869.00",0.00,0.03,0.00,0.00,0.00,0.00,1.00
CORRECTIONS,"4,362,869.00",0.01,0.13,0.00,0.00,0.00,0.00,84.00
CPV,"4,357,860.00","47,412,065.50","21,585,283.86","1,000,000.00","33,162,100.00","39,160,000.00","66,510,000.00","99,100,000.00"
MAIN_CPV_CODE_GPA,"1,639,715.00",109.88,203.69,0.00,33.00,39.00,71.00,806.00
GPA_COVERAGE,"1,639,715.00",1.94,1.34,1.00,1.00,1.00,3.00,5.00
LOTS_NUMBER,"4,350,004.00",21.66,76.01,0.00,0.00,1.00,11.00,"1,137.00"
VALUE_EURO,"2,624,436.00","524,784,487,241,368,862,720.00","831,779,254,787,191,208,935,424.00",0.00,"203,058.14","535,109.90","1,799,187.00","1,347,157,649,092,480,597,193,916,416.00"


#### Notes: Descriptype analyse of Numerical Data
1. General structure of numeric variables
Most numeric fields have very large counts (3–4 million), confirming that the dataset is broadly complete for core procurement indicators.
Several variables show substantial variance and heavy right‑skew (e.g., award values, estimated values), which is typical for procurement datasets where a small number of tenders have extremely high contract values.

2. Variables with strong completeness (recommended to keep)
These fields have high counts and stable distributions:
- YEAR — fully populated; essential for trend analysis.
- ID_NOTICE_CAN, ID_AWARD, ID_TYPE — identifiers with complete coverage.
- NUMBER_OFFERS — key competition indicator; distribution shows realistic variation (min=0, max>100).
- VALUE_EURO, AWARD_VALUE_EURO, AWARD_EST_VALUE_EURO — central economic variables with high coverage.
- LOTS_NUMBER — well‑populated; useful for structural analysis of tenders.
- NUMBER_AWARDS — consistent and complete.
These variables form the backbone of quantitative analysis.

3. Variables with partial completeness (use with caution)
Some fields have moderate missingness or inconsistent reporting:
- NUMBER_TENDERS_SME
- NUMBER_TENDERS_OTHER_EU
- NUMBER_TENDERS_NON_EU
- NUMBER_OFFERS_ELECTR
These variables are often sparsely populated across countries and years.
They may be retained for descriptive analysis but are not reliable for modeling.

4. Variables with extremely high missingness (recommended for removal)
The following numeric fields show very low counts relative to the dataset size:
- VALUE_EURO_FIN_1
- VALUE_EURO_FIN_2
- AWARD_VALUE_EURO_FIN_1
These appear to be alternative or secondary financial fields, often duplicating the main award value or containing incomplete reporting.
Given their low coverage and redundancy, they can be safely dropped.

5. Distributional observations
- Contract values (VALUE_EURO, AWARD_VALUE_EURO)
  - Highly skewed with very large maximum values (tens or hundreds of millions).
  - Log‑transformation is recommended for modeling.
- NUMBER_OFFERS
  - Median around 3–4 offers, but minimum is 0 and maximum exceeds 100.
  - This confirms the presence of both failed tenders (0–1 offers) and highly competitive ones.
- LOTS_NUMBER
  - Median is 1, but maximum is very high, indicating multi‑lot mega‑projects.
  - Consider capping or binning for modeling.

6. Recommended preprocessing steps based on describe()
Convert financial fields to float and handle outliers (log-transform or winsorize).
Convert date fields to datetime (DT_AWARD, DT_DISPATCH, DT_RECEIPT).
Remove secondary financial fields with low coverage.
Check for unrealistic values (e.g., negative amounts, extremely large numbers).
Create derived features:
- IS_FAILED_TENDER (0–1 offers)
- IS_LOW_COMPETITION (2 offers)
- LOG_AWARD_VALUE
- CPV_SECTION (first 2 digits of CPV)

---------------------
### Dublicats

---------------

In [11]:
df.duplicated().sum()

np.int64(0)

In [12]:
df.drop(columns=["YEAR"]).duplicated().sum()

np.int64(0)

#### Notes: Duplicate Check Summary
The dataset contains no full duplicates, even when the YEAR column is excluded from comparison.
Additionally, there are no repeated tenders with identical characteristics across different years, indicating that each procurement record represents a unique publication without cross‑year duplication.

--------------------

### ANALYSE TARGET

-----------------------

In [13]:
#-----------------------------
# Analyse for European Union
#-----------------------------

total = len(df)

failed = df[df["NUMBER_OFFERS"].fillna(0) <= 1]
risk = df[df["NUMBER_OFFERS"] == 2]

print("European Union:")
print("Total number of tenders in EU:", len(df))
print("Failed tenders (0–1 participant)):", len(failed))
print("Share (in %) of the total:", round(len(failed) / total * 100, 2), "%")

print("High-risk tenders (2 participants):", len(risk))
print("Share (in %) of the total:", round(len(risk) / total * 100, 2), "%")

European Union:
Total number of tenders in EU: 4362869
Failed tenders (0–1 participant)): 1676668
Share (in %) of the total: 38.43 %
High-risk tenders (2 participants): 576874
Share (in %) of the total: 13.22 %


In [14]:
#-----------------------------
# Analyse for Germany
#-----------------------------

# filter by Germany only
df_de = filter_germany(df)

# Competition check

total = len(df_de)

failed = df_de[df_de["NUMBER_OFFERS"].fillna(0) <= 1]
risk = df_de[df_de["NUMBER_OFFERS"] == 2]

print("Germany:")
print("Number of tenders in Germany:", len(df))
print("Failed tenders (0–1 participant)):", len(failed))
print("Share (in %) of the total:", round(len(failed) / total * 100, 2), "%")

print("High-risk tenders (2 participants):", len(risk))
print("Share (in %) of the total:", round(len(risk) / total * 100, 2), "%")



Germany:
Number of tenders in Germany: 4362869
Failed tenders (0–1 participant)): 84747
Share (in %) of the total: 27.37 %
High-risk tenders (2 participants): 33235
Share (in %) of the total: 10.73 %


#### Notes:  Dataset Patterns

1. Overall EU

The full dataset contains 4.36 million procurement records, providing a broad and diverse basis for statistical analysis and model training.
The share of failed tenders (0–1 participant) is 38.43%, indicating a substantial portion of procedures with insufficient competition.
High‑risk tenders (2 participants) represent 13.22% of all records, suggesting that limited competition is a recurring structural pattern across the EU market.
These proportions highlight consistent cross‑country patterns of low bidder participation, which can be leveraged for robust modeling and comparative analysis.

2. Germany‑Specific Patterns

Germany accounts for 309,652 records, a sufficiently large subset for targeted descriptive analysis but comparatively smaller for standalone predictive modeling.
The share of failed tenders in Germany is 27.37%, notably lower than the EU average, indicating relatively stronger competitive dynamics.
High‑risk tenders represent 10.73%, also below the EU benchmark, suggesting a more stable procurement environment with fewer structurally risky procedures.
The deviation of German indicators from EU‑wide averages implies country‑specific procurement behavior that merits separate examination.

-------------------------

### SAVE DATASET

----------------------

In [ ]:
# df.to_pickle("data/dataset.pkl")
# df_de.to_pickle("data/dataset_de.pkl")